cpp/src/spatial/linestring_distance.cu

```diff
@@ -0,0 +1,334 @@
+void __global__ pairwise_linestring_distance_kernel(OffsetIterator linestring1_offsets_begin,
+                                                    OffsetIterator linestring1_offsets_end,
+                                                    CoordinateIterator linestring1_points_xs_begin,
+                                                    CoordinateIterator linestring1_points_xs_end,
+                                                    CoordinateIterator linestring1_points_ys_begin,
+                                                    OffsetIterator linestring2_offsets_begin,
+                                                    CoordinateIterator linestring2_points_xs_begin,
+                                                    CoordinateIterator linestring2_points_xs_end,
+                                                    CoordinateIterator linestring2_points_ys_begin,
+                                                    OutputIterator distances)
+{
+  using T = typename std::iterator_traits<CoordinateIterator>::value_type;
+
+  auto const p1_idx = threadIdx.x + blockIdx.x * blockDim.x;
+  cudf::size_type const num_linestrings =
+    thrust::distance(linestring1_offsets_begin, linestring1_offsets_end);
+  cudf::size_type const linestring1_num_points =
+    thrust::distance(linestring1_points_xs_begin, linestring1_points_xs_end);
+  cudf::size_type const linestring2_num_points =
+    thrust::distance(linestring2_points_xs_begin, linestring2_points_xs_end);
+
+  if (p1_idx >= linestring1_num_points) { return; }
+
+  cudf::size_type const linestring_idx =
+    thrust::distance(linestring1_offsets_begin,
+                     thrust::upper_bound(
+                       thrust::seq, linestring1_offsets_begin, linestring1_offsets_end, p1_idx)) -
+    1;
+
+  cudf::size_type ls1_end = endpoint_index_of_linestring(
+    linestring_idx, linestring1_offsets_begin, num_linestrings, linestring1_num_points);
+
+  if (p1_idx == ls1_end) {
+    // Current point is the end point of the line string.
+    return;
+  }
+
+  cudf::size_type ls2_start = *(linestring2_offsets_begin + linestring_idx);
+  cudf::size_type ls2_end   = endpoint_index_of_linestring(
+    linestring_idx, linestring2_offsets_begin, num_linestrings, linestring2_num_points);
+
+  vec_2d<T> A{linestring1_points_xs_begin[p1_idx], linestring1_points_ys_begin[p1_idx]};
+  vec_2d<T> B{linestring1_points_xs_begin[p1_idx + 1], linestring1_points_ys_begin[p1_idx + 1]};
+
+  T min_distance = std::numeric_limits<T>::max();
+  for (cudf::size_type p2_idx = ls2_start; p2_idx < ls2_end; p2_idx++) {
+    vec_2d<T> C{linestring2_points_xs_begin[p2_idx], linestring2_points_ys_begin[p2_idx]};
+    vec_2d<T> D{linestring2_points_xs_begin[p2_idx + 1], linestring2_points_ys_begin[p2_idx + 1]};
+    min_distance = std::min(min_distance, segment_distance(A, B, C, D));
+  }
+  atomicMin(distances + linestring_idx, static_cast<T>(min_distance));
```

Comment:
@harrism mentioned this [here](https://github.com/rapidsai/cuspatial/pull/510#pullrequestreview-954178500), but we should be able to remove the `std::sqrt()` from `segment_distance`, comparing squared distances and taking the root of the min squared distance at the end (I recall seeing a nice performance improvement when we did this in [`point_to_poly_line_distance`](https://github.com/rapidsai/cuspatial/blob/1986a18e529888f9fe8df7c120924fe52cb1f06e/cpp/src/utility/point_to_nearest_polyline.cuh#L57-L60)):

```suggestion
  atomicMin(distances + linestring_idx, static_cast<T>(std::sqrt(min_distance)));
```

Reply:
Good catch! I thought @isVoid had already made this change, I didn't look close enough.

Reply:
I've renamed the variable (and method) with `squared_`.